{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "RUx-jOSj6qql",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import json\n",
    "import urllib\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib.patches import Polygon\n",
    "from scipy import stats\n",
    "\n",
    "import dataParser as parser\n",
    "import helperFiles.buildPlot as plotBuilder\n",
    "import pickle\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "Swgj940_6qqp",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv('../merged_1.6.1.csv')\n",
    "\n",
    "data = []\n",
    "labels = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 54
    },
    "colab_type": "code",
    "id": "IMlPj2pa6qqs",
    "outputId": "6d25cdff-a1e8-4e3a-d4e7-83c63b32ebbb",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Nonhomologous end-joining': ['XRCC6', 'XRCC5', 'PRKDC', 'LIG4', 'XRCC4', 'DCLRE1C', 'NHEJ1'], 'Microhomology end-joining': ['MRE11', 'RAD50', 'NBN', 'RBBP8', 'ERCC4', 'ERCC1', 'LIG1', 'POLL', 'POLB', 'PARP1', 'LIG3', 'XRCC1'], 'Homologous recombination': ['RAD51', 'RAD51B', 'RAD51D', 'DMC1', 'XRCC2', 'XRCC3', 'RAD52', 'RAD54L', 'RAD54B', 'BRCA1', 'SHFM1', 'RAD50', 'MRE11', 'NBN', 'RBBP8', 'MUS81', 'EME1', 'EME2', 'GIYD1', 'GIYD2', 'GEN1'], 'Base excision repair': ['UNG', 'SMUG1', 'MBD4', 'TDG', 'OGG1', 'MUTYH', 'NTHL1', 'MPG', 'NEIL1', 'NEIL2', 'NEIL3', 'APEX1', 'APEX2', 'LIG3', 'XRCC1', 'PNKP', 'APLF', 'PARP1', 'PARP2', 'PARP3', 'MGMT', 'ALKBH2', 'ALKBH3'], 'Repair of DNA-topoisomerase crosslinks': ['TDP1', 'TDP2'], 'Mismatch excision repair': ['MSH2', 'MSH3', 'MSH6', 'MLH1', 'PMS2', 'MSH4', 'MSH5', 'MLH3', 'PMS1', 'PMS2L3'], 'Nucleotide excision repair': ['RAD23B', 'CETN2', 'RAD23A', 'XPA', 'DDB1', 'DDB2', 'RPA1', 'RPA2', 'RPA3', 'TFIIH', 'ERCC3', 'ERCC2', 'GTF2H1', 'GTF2H2', 'GTF2H3', 'GTF2H4', 'GTF2H5', 'CDK7', 'CCNH', 'MNAT1', 'ERCC5', 'ERCC1', 'ERCC4', 'LIG1', 'ERCC8', 'ERCC6', 'UVSSA', 'XAB2', 'MMS19'], 'Fanconi anemia': ['FANCA', 'FANCB', 'FANCC', 'BRCA2', 'FANCD2', 'FANCE', 'FANCF', 'FANCG', 'FANCI', 'BRIP1', 'FANCL', 'FANCM', 'PALB2', 'RAD51C', 'BTBD12', 'FAAP20', 'FAAP24'], 'Modulation of nucleotide pools': ['NUDT1', 'DUT', 'RRM2B'], 'DNA polymerases': ['POLB', 'POLG', 'POLD1', 'POLE', 'PCNA', 'REV3L', 'MAD2L2', 'REV1L', 'POLH', 'POLI', 'POLQ', 'POLK', 'POLL', 'POLM', 'POLN'], 'Editing and processing nucleases': ['FEN1', 'FAN1', 'TREX1', 'TREX2', 'EXO1', 'APTX', 'SPO11', 'ENDOV'], 'Ubiquitination and modification': ['UBE2A', 'UBE2B', 'RAD18', 'SHPRH', 'HLTF', 'RNF168', 'SPRTN', 'RNF8', 'RNF4', 'UBE2V2', 'UBE2N'], 'Chromatin Structure and Modification': ['H2AFX', 'CHAF1A', 'SETMAR'], 'Other conserved DNA damage response genes': ['ATR', 'ATRIP', 'MDC1', 'RAD1', 'RAD9A', 'HUS1', 'RAD17', 'CHEK1', 'CHEK2', 'TP53', 'TP53BP1', 'RIF1', 'TOPBP1', 'CLK2', 'PER1']}\n"
     ]
    }
   ],
   "source": [
    "print(parser.samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 54
    },
    "colab_type": "code",
    "id": "rprZ5CBH6qqv",
    "outputId": "682a0980-f113-42e4-ace2-42e611b3f786",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'TDP1', 'POLE', 'TREX2', 'MSH2', 'RBBP8', 'DDB1', 'FANCF', 'CHEK2', 'UBE2V2', 'RPA1', 'RAD50', 'FANCE', 'TDG', 'PER1', 'RAD1', 'XRCC1', 'ERCC2', 'RRM2B', 'POLD1', 'HLTF', 'PARP1', 'APLF', 'MAD2L2', 'POLQ', 'CLK2', 'RAD23A', 'LIG4', 'ERCC4', 'BRCA1', 'GTF2H5', 'CHEK1', 'GEN1', 'ALKBH2', 'PMS2', 'BRIP1', 'ERCC8', 'XRCC4', 'UNG', 'RAD54B', 'FANCB', 'TREX1', 'FANCM', 'MDC1', 'FANCC', 'FAAP24', 'MGMT', 'MSH4', 'RAD18', 'FEN1', 'POLM', 'RPA3', 'RAD51B', 'OGG1', 'ATR', 'SHFM1', 'ERCC5', 'SHPRH', 'PRKDC', 'NHEJ1', 'RAD51C', 'DCLRE1C', 'PARP3', 'POLI', 'XRCC5', 'FANCA', 'ERCC1', 'MPG', 'MRE11', 'REV1L', 'PNKP', 'BTBD12', 'FANCI', 'FANCG', 'RIF1', 'MNAT1', 'BRCA2', 'TOPBP1', 'HUS1', 'DMC1', 'RAD51', 'CDK7', 'TP53', 'GTF2H3', 'NTHL1', 'ERCC6', 'MSH6', 'H2AFX', 'POLB', 'PMS1', 'EXO1', 'LIG1', 'MMS19', 'XRCC6', 'DUT', 'RNF4', 'UBE2B', 'NBN', 'RAD23B', 'CCNH', 'RAD54L', 'GIYD1', 'RNF168', 'NUDT1', 'PMS2L3', 'FAN1', 'XRCC3', 'RNF8', 'XAB2', 'RAD51D', 'APTX', 'MLH1', 'FAAP20', 'SETMAR', 'RPA2', 'REV3L', 'MLH3', 'CETN2', 'POLK', 'POLL', 'RAD17', 'CHAF1A', 'SMUG1', 'TP53BP1', 'GTF2H2', 'GTF2H1', 'RAD9A', 'NEIL1', 'MUS81', 'RAD52', 'MBD4', 'UBE2N', 'ERCC3', 'FANCL', 'UVSSA', 'POLG', 'ALKBH3', 'MSH3', 'SPO11', 'PCNA', 'LIG3', 'EME2', 'ATRIP', 'MUTYH', 'NEIL2', 'GTF2H4', 'GIYD2', 'ENDOV', 'NEIL3', 'FANCD2', 'SPRTN', 'XRCC2', 'POLH', 'MSH5', 'EME1', 'POLN', 'TDP2', 'XPA', 'PARP2', 'TFIIH', 'PALB2', 'UBE2A', 'APEX2', 'DDB2', 'APEX1'}\n"
     ]
    }
   ],
   "source": [
    "inputs = set()\n",
    "samples = parser.samples\n",
    "for category in samples:\n",
    "    for gene in samples[category]:\n",
    "        inputs.add(gene)\n",
    "print(inputs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "colab_type": "code",
    "id": "Lh7TZMwq6qqz",
    "outputId": "50c0685e-a03b-444e-c87c-5a86641c8f69",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TDP1\n",
      "0.0005195028837750801\n",
      "POLE\n",
      "6.345111772667445e-09\n",
      "TREX2\n",
      "invalid literal for int() with base 10: 'X'\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/TREX2?content-type=application/json;expand=1\n",
      "symb got fricked\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/numpy/core/fromnumeric.py:3584: RuntimeWarning: Degrees of freedom <= 0 for slice\n",
      "  **kwargs)\n",
      "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/numpy/core/_methods.py:209: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ret = ret.dtype.type(ret / rcount)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nan\n",
      "MSH2\n",
      "6.101528754718389e-21\n",
      "RBBP8\n",
      "0.00010630509553453367\n",
      "DDB1\n",
      "0.25265642986297404\n",
      "FANCF\n",
      "nan\n",
      "CHEK2\n",
      "0.17370842405287698\n",
      "UBE2V2\n",
      "0.4715591615067086\n",
      "RPA1\n",
      "1.9943220099486137e-08\n",
      "RAD50\n",
      "0.004071881239984347\n",
      "FANCE\n",
      "0.09836791172071749\n",
      "TDG\n",
      "8.185378496501703e-15\n",
      "PER1\n",
      "0.41948011203981717\n",
      "RAD1\n",
      "0.5354893843015642\n",
      "XRCC1\n",
      "0.010336661231414466\n",
      "ERCC2\n",
      "0.12230264724107563\n",
      "RRM2B\n",
      "0.15258139943110818\n",
      "POLD1\n",
      "1.0509060424015153e-21\n",
      "HLTF\n",
      "0.0010088109205907889\n",
      "PARP1\n",
      "6.127376244576102e-14\n",
      "APLF\n",
      "3.6915751987558677e-10\n",
      "MAD2L2\n",
      "0.5198309595705277\n",
      "POLQ\n",
      "0.3856994080337962\n",
      "CLK2\n",
      "0.7548125835160282\n",
      "RAD23A\n",
      "0.9667185428758394\n",
      "LIG4\n",
      "nan\n",
      "ERCC4\n",
      "nan\n",
      "BRCA1\n",
      "1.2027544362577605e-10\n",
      "GTF2H5\n",
      "nan\n",
      "CHEK1\n",
      "nan\n",
      "GEN1\n",
      "0.0005091088304816632\n",
      "ALKBH2\n",
      "nan\n",
      "PMS2\n",
      "nan\n",
      "BRIP1\n",
      "5.075882791806269e-17\n",
      "ERCC8\n",
      "0.003567687788302094\n",
      "XRCC4\n",
      "2.946833567904031e-13\n",
      "UNG\n",
      "0.0002870462650760585\n",
      "RAD54B\n",
      "2.471269514322044e-06\n",
      "FANCB\n",
      "invalid literal for int() with base 10: 'X'\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/FANCB?content-type=application/json;expand=1\n",
      "symb got fricked\n",
      "nan\n",
      "TREX1\n",
      "nan\n",
      "FANCM\n",
      "0.0075863057414373375\n",
      "MDC1\n",
      "0.003117388889882213\n",
      "FANCC\n",
      "1.1864611240045823e-29\n",
      "FAAP24\n",
      "0.4536801997465062\n",
      "MGMT\n",
      "0.013523197082176627\n",
      "MSH4\n",
      "0.136316579040129\n",
      "RAD18\n",
      "1.656149590646894e-17\n",
      "FEN1\n",
      "0.5105469388053578\n",
      "POLM\n",
      "nan\n",
      "RPA3\n",
      "7.8557192170654415e-25\n",
      "RAD51B\n",
      "2.085113142152007e-42\n",
      "OGG1\n",
      "1.081443957674269e-10\n",
      "ATR\n",
      "0.00026518657607234634\n",
      "SHFM1\n",
      "HTTP Error 400: Bad Request\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/SHFM1?content-type=application/json;expand=1\n",
      "symb got fricked\n",
      "nan\n",
      "ERCC5\n",
      "0.0037875460250282214\n",
      "SHPRH\n",
      "3.1125256346063406e-05\n",
      "PRKDC\n",
      "1.2194936825568415e-09\n",
      "NHEJ1\n",
      "0.0005371086330100564\n",
      "RAD51C\n",
      "1.030990788504844e-12\n",
      "DCLRE1C\n",
      "1.1057408222510404e-15\n",
      "PARP3\n",
      "nan\n",
      "POLI\n",
      "0.6205590159763072\n",
      "XRCC5\n",
      "0.00010381054876323055\n",
      "FANCA\n",
      "6.348688907805628e-10\n",
      "ERCC1\n",
      "6.65841039653029e-08\n",
      "MPG\n",
      "nan\n",
      "MRE11\n",
      "0.0005255941859959409\n",
      "REV1L\n",
      "HTTP Error 400: Bad Request\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/REV1L?content-type=application/json;expand=1\n",
      "symb got fricked\n",
      "nan\n",
      "PNKP\n",
      "nan\n",
      "BTBD12\n",
      "HTTP Error 400: Bad Request\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/BTBD12?content-type=application/json;expand=1\n",
      "symb got fricked\n",
      "nan\n",
      "FANCI\n",
      "0.12379565852031262\n",
      "FANCG\n",
      "nan\n",
      "RIF1\n",
      "0.033054857915190826\n",
      "MNAT1\n",
      "9.112675047624477e-06\n",
      "BRCA2\n",
      "0.0007761167277485903\n",
      "TOPBP1\n",
      "2.220326243778787e-06\n",
      "HUS1\n",
      "nan\n",
      "DMC1\n",
      "6.057780830661484e-10\n",
      "RAD51\n",
      "0.6727184003648654\n",
      "CDK7\n",
      "nan\n",
      "TP53\n",
      "0.8675734941287405\n",
      "GTF2H3\n",
      "3.581528711820842e-06\n",
      "NTHL1\n",
      "0.06448706613895508\n",
      "ERCC6\n",
      "5.837198745372309e-05\n",
      "MSH6\n",
      "5.316781792651761e-13\n",
      "H2AFX\n",
      "HTTP Error 400: Bad Request\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/H2AFX?content-type=application/json;expand=1\n",
      "symb got fricked\n",
      "nan\n",
      "POLB\n",
      "0.004266016259467122\n",
      "PMS1\n",
      "7.226822547667796e-19\n",
      "EXO1\n",
      "2.917596494958559e-12\n",
      "LIG1\n",
      "0.0038560546421134863\n",
      "MMS19\n",
      "0.141867180752654\n",
      "XRCC6\n",
      "6.356398437160524e-12\n",
      "DUT\n",
      "nan\n",
      "RNF4\n",
      "0.00018574078957441415\n",
      "UBE2B\n",
      "0.48923747261020367\n",
      "NBN\n",
      "1.7067372155020075e-08\n",
      "RAD23B\n",
      "2.3435490270589447e-07\n",
      "CCNH\n",
      "0.0031255931946823747\n",
      "RAD54L\n",
      "0.0014816305880700697\n",
      "GIYD1\n",
      "HTTP Error 400: Bad Request\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/GIYD1?content-type=application/json;expand=1\n",
      "symb got fricked\n",
      "nan\n",
      "RNF168\n",
      "0.7203456547325313\n",
      "NUDT1\n",
      "nan\n",
      "PMS2L3\n",
      "HTTP Error 400: Bad Request\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/PMS2L3?content-type=application/json;expand=1\n",
      "symb got fricked\n",
      "nan\n",
      "FAN1\n",
      "nan\n",
      "XRCC3\n",
      "0.3530659082143972\n",
      "RNF8\n",
      "0.17000880148331093\n",
      "XAB2\n",
      "nan\n",
      "RAD51D\n",
      "0.12408178235007981\n",
      "APTX\n",
      "1.9450369573194032e-07\n",
      "MLH1\n",
      "0.9014820507439288\n",
      "FAAP20\n",
      "2.1456688447307003e-11\n",
      "SETMAR\n",
      "3.125180449408741e-05\n",
      "RPA2\n",
      "0.0027439327248136424\n",
      "REV3L\n",
      "4.255388452784343e-25\n",
      "MLH3\n",
      "0.009170037441057903\n",
      "CETN2\n",
      "invalid literal for int() with base 10: 'X'\n",
      "https://rest.ensembl.org/lookup/symbol/homo_sapiens/CETN2?content-type=application/json;expand=1\n",
      "symb got fricked\n",
      "nan\n",
      "POLK\n",
      "0.0005631484901593005\n",
      "POLL\n",
      "0.011370618257855562\n",
      "RAD17\n",
      "nan\n",
      "CHAF1A\n",
      "0.013727289823013822\n",
      "SMUG1\n",
      "3.306900792068522e-16\n",
      "TP53BP1\n",
      "0.10939997241233229\n",
      "GTF2H2\n",
      "nan\n",
      "GTF2H1\n",
      "0.0005241976635679014\n",
      "RAD9A\n",
      "1.2359113304059913e-14\n",
      "NEIL1\n",
      "nan\n",
      "MUS81\n",
      "0.0008897549296430293\n",
      "RAD52\n",
      "9.112298010487913e-06\n",
      "MBD4\n",
      "nan\n",
      "UBE2N\n",
      "5.379204397458186e-10\n",
      "ERCC3\n",
      "nan\n",
      "FANCL\n",
      "3.713669673038094e-10\n",
      "UVSSA\n",
      "0.00011647837817173482\n",
      "POLG\n",
      "0.6751341762628027\n",
      "ALKBH3\n",
      "5.16873821016746e-07\n",
      "MSH3\n",
      "3.934310528218806e-05\n",
      "SPO11\n",
      "3.079750999658559e-06\n",
      "PCNA\n",
      "0.9624179566741957\n",
      "LIG3\n",
      "0.00030687392596743184\n",
      "EME2\n",
      "5.0024573468317755e-05\n",
      "ATRIP\n"
     ]
    }
   ],
   "source": [
    "pValues = []\n",
    "for gene in inputs:\n",
    "    print(gene)\n",
    "    all_donors = df['donor_unique_id'].unique()\n",
    "    donorsInGene = parser.getDonors(gene)\n",
    "    diffDonors = parser.diff(all_donors, donorsInGene)\n",
    "\n",
    "    affected = []\n",
    "    notaffected = []\n",
    "\n",
    "    for donor in donorsInGene:\n",
    "        donorDF = df[(df['donor_unique_id'] == donor)]\n",
    "        affected.append(len(donorDF.index))\n",
    "\n",
    "    for donor in diffDonors:\n",
    "        donorDF = df[(df['donor_unique_id'] == donor)]\n",
    "        notaffected.append(len(donorDF.index))\n",
    "    if(len(affected) == 0):\n",
    "        affected.append(0)\n",
    "    if(len(notaffected) == 0):\n",
    "        notaffected.append(0)\n",
    "    data.append(notaffected)\n",
    "    data.append(affected)\n",
    "    labels.append(\"WT \" + gene)\n",
    "    labels.append(\"MUT \" + gene)\n",
    "    value = stats.ttest_ind(affected, notaffected)[1]\n",
    "    pValues.append(value)\n",
    "    print(value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 340
    },
    "colab_type": "code",
    "id": "7aESVwMNGwM_",
    "outputId": "e0a027af-51fe-41ff-b5f0-443e5b7c2ac0",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "parser.save_obj(pValues,'pvalues')\n",
    "data = load_obj(data, 'data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "_7w7IHAQG8lY",
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# import mpld3\n",
    "# mpld3.enable_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "colab_type": "code",
    "id": "Wk6nalGiHDRi",
    "outputId": "b13adffe-f55e-4fb8-a1a3-36293e16a9f9",
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig, ax1 = plt.subplots(figsize=(15, 5))\n",
    "fig.canvas.set_window_title('A Boxplot Example')\n",
    "fig.subplots_adjust(left=0.075, right=0.95, top=0.9, bottom=0.25)\n",
    "\n",
    "bp = ax1.boxplot(data, notch=0, sym='+', vert=1, whis=1.5, showmeans=True)\n",
    "plt.setp(bp['boxes'], color='black')\n",
    "plt.setp(bp['whiskers'], color='black')\n",
    "plt.setp(bp['fliers'], color='red', marker='+')\n",
    "\n",
    "# Add a horizontal grid to the plot, but make it very light in color\n",
    "# so we can use it for reading data values but not be distracting\n",
    "ax1.yaxis.grid(True, linestyle='-', which='major', color='lightgrey',\n",
    "               alpha=0.5)\n",
    "\n",
    "# Hide these grid behind plot objects\n",
    "ax1.set_axisbelow(True)\n",
    "ax1.set_title(\n",
    "    'Number of SVs Per Donor In Regular Gene vs Mutated')\n",
    "ax1.set_xlabel('Gene')\n",
    "ax1.set_ylabel('Number of SVs Per Donor')\n",
    "\n",
    "# Now fill the boxes with desired colors\n",
    "box_colors = ['darkkhaki', 'royalblue']\n",
    "num_boxes = len(data)\n",
    "medians = np.empty(num_boxes)\n",
    "for i in range(num_boxes):\n",
    "    box = bp['boxes'][i]\n",
    "    boxX = []\n",
    "    boxY = []\n",
    "    for j in range(5):\n",
    "        boxX.append(box.get_xdata()[j])\n",
    "        boxY.append(box.get_ydata()[j])\n",
    "    box_coords = np.column_stack([boxX, boxY])\n",
    "    # Alternate between Dark Khaki and Royal Blue\n",
    "    ax1.add_patch(Polygon(box_coords, facecolor=box_colors[i % 2]))\n",
    "    # Now draw the median lines back over what we just filled in\n",
    "    med = bp['medians'][i]\n",
    "    medianX = []\n",
    "    medianY = []\n",
    "    for j in range(2):\n",
    "        medianX.append(med.get_xdata()[j])\n",
    "        medianY.append(med.get_ydata()[j])\n",
    "        ax1.plot(medianX, medianY, 'k')\n",
    "    medians[i] = medianY[0]\n",
    "    # Finally, overplot the sample averages, with horizontal alignment\n",
    "    # in the center of each box\n",
    "    ax1.plot(np.average(med.get_xdata()), np.average(data[i]),\n",
    "             color='w', marker='*', markeredgecolor='k')\n",
    "\n",
    "# Set the axes ranges and axes labels\n",
    "ax1.set_xlim(0.5, num_boxes + 0.5)\n",
    "top = 2000\n",
    "bottom = 0\n",
    "ax1.set_ylim(bottom, top)\n",
    "ax1.set_xticklabels(labels,\n",
    "                    rotation=90, fontsize=8)\n",
    "\n",
    "# Due to the Y-axis scale being different across samples, it can be\n",
    "# hard to compare differences in medians across the samples. Add upper\n",
    "# X-axis tick labels with the sample medians to aid in comparison\n",
    "# (just use two decimal places of precision)\n",
    "pos = np.arange(num_boxes) + 1\n",
    "upper_labels = [str(np.round(s, 2)) for s in medians]\n",
    "weights = ['bold', 'semibold']\n",
    "for tick, label in zip(range(num_boxes), ax1.get_xticklabels()):\n",
    "    k = tick % 2\n",
    "    ax1.text(pos[tick], .95, upper_labels[tick],\n",
    "             transform=ax1.get_xaxis_transform(),\n",
    "             horizontalalignment='center', size='x-small',\n",
    "             weight=weights[k], color=box_colors[k])\n",
    "\n",
    "# Finally, add a basic legend\n",
    "fig.text(0.80, 0.08, f'Regular Gene',\n",
    "         backgroundcolor=box_colors[0], color='black', weight='roman',\n",
    "         size='x-small')\n",
    "fig.text(0.80, 0.045, 'Mutated Gene',\n",
    "         backgroundcolor=box_colors[1],\n",
    "         color='white', weight='roman', size='x-small')\n",
    "fig.text(0.80, 0.015, '*', color='white', backgroundcolor='silver',\n",
    "         weight='roman', size='medium')\n",
    "fig.text(0.815, 0.013, ' Average Value', color='black', weight='roman',\n",
    "         size='x-small')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "DBPs6nqzHQhU",
    "scrolled": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "name": "BarplotWithDifferencesOfWholeGenomeSVs.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
